zerver/lib/narrow.py

```diff
@@ -861,6 +873,65 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+# This function corrects the conversation filters of the narrow - channel,
+# topic, DM conversation to point to the narrow in which the message
+# id operand of the `with` operator is present in.
+#
+# This is done as an effort to facilitate permalinks for topics -- if there
+# exists a message represented by the id (`with` operand), we redirect to
+# the narrow in which the message is present. Else, we fallback
+# to the narrow mentioned originally.
+def update_narrow_terms_containing_with_operator(
+    realm: Realm,
+    maybe_user_profile: Union[UserProfile, AnonymousUser],
+    narrow: Optional[List[NarrowParameter]],
+) -> Optional[List[NarrowParameter]]:
+    if narrow is None:
+        return narrow
+
+    contains_with_operator = list(filter(lambda term: term.operator == "with", narrow))
```

Comment:
Is there some code that rejects having multiple `with` operators in a single request? Seems like we should do that and have a test that verifies we do that.

Reply:
updated so that in request with multiple `with` operators, narrow without any `with` operator would be returned.

Reply:
That seems very odd, why shouldn't that just be an error, like other invalid operator combinations?